Fix msvc compiler error C4146 (minus operator applied to unsigned type)

kobake opened the pull request:
On msvc14, the error C4146 (unary minus operator applied to unsigned type, result still unsigned) had been occured.
Minus operator should be applied to signed type.

kobake committed:
Fix msvc compiler error C4146 (minus operator applied to unsigned type)

On msvc14, the compiler error C4146 (unary minus operator applied to unsigned type, result still unsigned) had been occured.
Use '0 - x' styled formula instead of '-x' so as to fix the error.

laanwj commented on src/bloom.cpp:
```diff
@@ -254,8 +254,8 @@ void CRollingBloomFilter::insert(const std::vector<unsigned char>& vKey)
-        uint64_t nGenerationMask1 = -(uint64_t)(nGeneration & 1);
-        uint64_t nGenerationMask2 = -(uint64_t)(nGeneration >> 1);
+        uint64_t nGenerationMask1 = (uint64_t)(-(int64_t)(nGeneration & 1));
+        uint64_t nGenerationMask2 = (uint64_t)(-(int64_t)(nGeneration >> 1));
```
This works around the error, but I don't think it's very elegant. What about just formulating it explicitly:
```
        uint64_t nGenerationMask1 = ~(uint64_t)(nGeneration & 1) + 1;
        uint64_t nGenerationMask2 = ~(uint64_t)(nGeneration >> 1) + 1;
```

laanwj replied:
Or even just:
```
        uint64_t nGenerationMask1 = 0 - (uint64_t)(nGeneration & 1);
        uint64_t nGenerationMask2 = 0 - (uint64_t)(nGeneration >> 1);
```
Not sure this is any more readable than the above though :)

kobake replied:
Thanks. I confirmed three ways work well as same in my local msvc14.
The latter you show is simple but not feel as explicit bit calculation.
I like the former that is readable clearly as two's complement formula, then I'll adopt it.

kobake replied:
Applied. (commit --amend && push -f)

gmaxwell commented:
ugh. MSVC produces the most absurd errors for perfectly valid, strictly conforming code. I don't have a problem with silencing it in this way that the code currently is (though IMO it makes it less clear).

The prior fix of casting to an integer can actually result in undefined behavior (at least generally if not actually in our case due to the sizes we use). a=INT_MIN; a = -a; is undefined. (highlighting the foolishness of MSFT's war on bit manipulation. :P -- as someone trying to address this warning in an 'obvious way' would potentially introduce undefined behavior as a result!)

laanwj commented:
Agreed, that this is necessary at all is silly. There's no ambiguity here. Unary minus -x equivalent to 0-x which is well-defined also for unsigned values which are defined to wrap around.

kobake commented:
Yes, we should take care not to meet undefined behavior.
Now in this case, even if nGeneration has INT_MIN value, it's no problem because the target values to be applied two's complement formula are "(nGeneration & 1)" or "(nGeneration >> 1)"; they are 32bit values and their most significant bit is always zero, of course it's true even after the cast.

gmaxwell commented:
https://docs.microsoft.com/en-us/cpp/error-messages/compiler-warnings/compiler-warning-level-2-c4146

Seems to indicate that 0-(uint64_t)(nGeneration & 1); will silence it-- if so I think that would be preferable.

kobake commented:
OK, I just understood. I'll adopt ```0 - x``` style.
FYI, msvc14 (vs2015) treats C4146 not as warning but as error by default because of ```/sdl``` option.

laanwj commented:
utACK https://github.com/bitcoin/bitcoin/pull/9916/commits/40aca33add8f108caa517a952f6c8555a63f6f17

laanwj commented:
Amended the commit message to put a newline between the title and body, otherwise `git log` in short formats doesn't work as expected.

kobake commented:
Oh, I see. I've inserted new line as you say. Thanks for helpful tips.

kobake committed:
Fix msvc compiler error C4146 (unary minus operator applied to unsigned type)

On msvc14, int literal '-2147483648' is invalid, because '2147483648' is unsigned type and cant't apply minus operator to unsigned type.
To define the int literal correctly, use '-2147483647 - 1' formula that is also used to define INT_MIN in limits.h.

kobake commented:
I also fixed another C4146 error in the commit below that is about unit test.
https://github.com/kobake/bitcoin-msvc/commit/3c73d820e2355b36ea6bd6c4b78e4afa82637b8f

Should I contain the commit in this PR #9916, or create an another PR for the commit?

laanwj commented:
It's another minor thing, I'd say add it here.

kobake commented:
Thanks, I added.

laanwj commented:
re-UTack https://github.com/bitcoin/bitcoin/pull/9916/commits/8e0720bdb9141b00b4c00893b8139904c67ddacf

paveljanik commented:
utACK https://github.com/bitcoin/bitcoin/pull/9916/commits/8e0720bdb9141b00b4c00893b8139904c67ddacf
